File: backend/src/routes/budget.ts
import express, { Response } from 'express';
import { docClient, TABLES, GetCommand, UpdateCommand, QueryCommand, getCurrentTimestamp } from '../db/dynamodb';
import { authenticateToken, AuthRequest } from '../middleware/auth';
import { BudgetSetting, Expense, RecurringCost } from '../types';

const router = express.Router();

// Get budget settings
router.get('/', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const result = await docClient.send(new GetCommand({
      TableName: TABLES.BUDGET_SETTINGS,
      Key: {
        user_id: req.userId,
      },
    }));

    if (!result.Item) {
      return res.status(404).json({ error: 'Budget settings not found' });
    }

    res.json(result.Item);
  } catch (error) {
    console.error('Get budget settings error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Update budget settings
router.put('/', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const { monthly_limit } = req.body;

    if (monthly_limit === undefined || monthly_limit === null) {
      return res.status(400).json({ error: 'Monthly limit is required' });
    }

    const result = await docClient.send(new UpdateCommand({
      TableName: TABLES.BUDGET_SETTINGS,
      Key: {
        user_id: req.userId,
      },
      UpdateExpression: 'SET monthly_limit = :limit, updated_at = :timestamp',
      ExpressionAttributeValues: {
        ':limit': monthly_limit,
        ':timestamp': getCurrentTimestamp(),
      },
      ReturnValues: 'ALL_NEW',
    }));

    res.json(result.Attributes);
  } catch (error) {
    console.error('Update budget settings error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get spending summary for a month
router.get('/summary/:year/:month', authenticateToken, async (req: AuthRequest, res: Response) => {
  try {
    const { year, month } = req.params;
    const paddedMonth = month.padStart(2, '0');
    const monthStart = `${year}-${paddedMonth}-01`;

    // Get all expenses for the user
    const expensesResult = await docClient.send(new QueryCommand({
      TableName: TABLES.EXPENSES,
      KeyConditionExpression: 'user_id = :userId',
      ExpressionAttributeValues: {
        ':userId': req.userId,
      },
    }));

    const expenses = (expensesResult.Items || []) as Expense[];

    // Filter expenses for the specific month and calculate total
    const monthExpenses = expenses.filter(expense => {
      const expenseDate = new Date(expense.date);
      return expenseDate.getFullYear() === parseInt(year) &&
             expenseDate.getMonth() === parseInt(month) - 1;
    });

    const totalSpent = monthExpenses.reduce((sum, exp) => sum + exp.amount, 0);
    const transactionCount = monthExpenses.length;

    // Get recurring costs
    const recurringResult = await docClient.send(new QueryCommand({
      TableName: TABLES.RECURRING_COSTS,
      KeyConditionExpression: 'user_id = :userId',
      ExpressionAttributeValues: {
        ':userId': req.userId,
      },
    }));

    const recurringCosts = (recurringResult.Items || []) as RecurringCost[];

    // Filter recurring costs that started before or during this month
    const activeRecurringCosts = recurringCosts.filter(cost => {
      const startDate = new Date(cost.start_date);
      const targetDate = new Date(monthStart);
      return startDate <= targetDate;
    });

    // Calculate recurring costs totals
    const recurringMonthly = activeRecurringCosts
      .filter(cost => cost.frequency === 'monthly')
      .reduce((sum, cost) => sum + cost.amount, 0);

    const recurringAnnual = activeRecurringCosts
      .filter(cost => cost.frequency === 'annual')
      .reduce((sum, cost) => sum + cost.amount, 0);

    // Get budget settings
    const budgetResult = await docClient.send(new GetCommand({
      TableName: TABLES.BUDGET_SETTINGS,
      Key: {
        user_id: req.userId,
      },
    }));

    const budgetSetting = budgetResult.Item as BudgetSetting;

    // Calculate summary
    const totalRecurring = recurringMonthly + (recurringAnnual / 12);
    const totalWithRecurring = totalSpent + totalRecurring;
    const remaining = budgetSetting.monthly_limit - totalWithRecurring;
    const percentageUsed = budgetSetting.monthly_limit > 0
      ? (totalWithRecurring / budgetSetting.monthly_limit) * 100
      : 0;

    res.json({
      total_spent: totalSpent,
      recurring_costs: totalRecurring,
      total_with_recurring: totalWithRecurring,
      budget_limit: budgetSetting.monthly_limit,
      remaining,
      percentage_used: percentageUsed,
      transaction_count: transactionCount,
      is_over_budget: totalWithRecurring > budgetSetting.monthly_limit
    });
  } catch (error) {
    console.error('Get spending summary error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/api/budget.py
from fastapi import APIRouter, HTTPException, Depends
from datetime import datetime
from typing import Dict, Any
from .models import BudgetSettings, BudgetResponse
from .database import (
    get_budget_settings,
    save_budget_settings,
    get_expenses_by_user,
    get_recurring_costs_by_user,
    get_current_timestamp
)
from .middleware import get_current_user

router = APIRouter(prefix="/budget", tags=["budget"])


@router.get("/settings", response_model=BudgetResponse)
async def get_budget(current_user: dict = Depends(get_current_user)):
    """Get budget settings for the authenticated user"""
    user_id = current_user['user_id']
    budget = get_budget_settings(user_id)

    if not budget:
        raise HTTPException(status_code=404, detail="Budget settings not found")

    return budget


@router.post("/settings", response_model=BudgetResponse)
async def update_budget(
    budget_data: BudgetSettings,
    current_user: dict = Depends(get_current_user)
):
    """Update budget settings"""
    user_id = current_user['user_id']

    budget = {
        'user_id': user_id,
        'monthly_budget': budget_data.monthly_budget,
        'updated_at': get_current_timestamp()
    }

    return save_budget_settings(budget)


@router.get("/spending/{year}/{month}")
async def get_spending_summary(
    year: int,
    month: int,
    current_user: dict = Depends(get_current_user)
) -> Dict[str, Any]:
    """Get spending summary for a specific month"""
    user_id = current_user['user_id']

    # Get all expenses for the user
    expenses = get_expenses_by_user(user_id)

    # Filter expenses for the specified month
    month_expenses = []
    for expense in expenses:
        expense_date = datetime.fromisoformat(expense['date'].replace('Z', '+00:00'))
        if expense_date.year == year and expense_date.month == month:
            month_expenses.append(expense)

    # Calculate total spending
    total_spent = sum(exp['amount'] for exp in month_expenses)

    # Get budget settings
    budget = get_budget_settings(user_id)
    monthly_budget = budget['monthly_budget'] if budget else 0

    # Calculate category breakdown
    category_totals = {}
    for expense in month_expenses:
        category = expense['category']
        category_totals[category] = category_totals.get(category, 0) + expense['amount']

    # Get recurring costs
    recurring_costs = get_recurring_costs_by_user(user_id)

    # Calculate monthly recurring total
    monthly_recurring = 0
    for cost in recurring_costs:
        if cost['frequency'] == 'monthly':
            monthly_recurring += cost['amount']
        elif cost['frequency'] == 'annual':
            monthly_recurring += cost['amount'] / 12

    # Calculate daily spending for chart
    daily_spending = {}
    for expense in month_expenses:
        expense_date = datetime.fromisoformat(expense['date'].replace('Z', '+00:00'))
        day = expense_date.day
        daily_spending[day] = daily_spending.get(day, 0) + expense['amount']

    # Build daily data for chart
    import calendar
    days_in_month = calendar.monthrange(year, month)[1]
    daily_data = []
    for day in range(1, days_in_month + 1):
        daily_data.append({
            'day': day,
            'amount': daily_spending.get(day, 0)
        })

    return {
        'total_spent': total_spent,
        'monthly_budget': monthly_budget,
        'remaining': monthly_budget - total_spent if monthly_budget else 0,
        'percentage_used': (total_spent / monthly_budget * 100) if monthly_budget > 0 else 0,
        'category_breakdown': category_totals,
        'monthly_recurring': monthly_recurring,
        'expense_count': len(month_expenses),
        'daily_spending': daily_data,
        'expenses': month_expenses,
        'recurring_costs': recurring_costs
    }
